tenacity/retry.py

```diff
@@ -30,10 +30,16 @@ def __call__(self, retry_state: "RetryCallState") -> bool:
         pass
 
     def __and__(self, other: "retry_base") -> "retry_all":
-        return retry_all(self, other)
+        return other.__rand__(self)
+
+    def __rand__(self, other: "retry_base") -> "retry_all":
+        return retry_all(other, self)
 
     def __or__(self, other: "retry_base") -> "retry_any":
-        return retry_any(self, other)
+        return other.__ror__(self)
+
+    def __ror__(self, other: "retry_base") -> "retry_any":
+        return retry_any(other, self)
 
```

Comment:
Could/should be part of a different PR AFAIU.

Reply:
Not really, as we are introducing new async strategies, if sync and async strategies got mixed then the process would fail, as async strategies [would not be awaited](https://github.com/jd/tenacity/blob/cb15300d9b4358d9bdb6cfd33d433368700b3abe/tenacity/retry.py#L266). This is the bit I had to introduce to allow that mix, without this [these tests don't pass](https://github.com/jd/tenacity/pull/451/files#diff-55234053c6f9538eec8b76e6fce1715585c5359a3a8362042d1041381c6bd751R258-R364).

Reply:
I'm not saying it should be merged after, but before as it could be useful for even current tenacity code, no?

Reply:
Ah I see. What would you like to see then, just these changes for `__rand__` and `__ror__` in a separate PR? I can do that, it just didn't feel that useful because as the codebase is now, there's only ever 1 type of `retry_base` and no `async` calls anywhere, so there won't be any undesired mix that makes these operations useful. It's in this PR that I'm introducing new `async` strategies that need these to avoid errors when mixing both strategies. Or maybe I am missing your point entirely and it's something else you are expecting 😅

Reply:
@jd any thoughts on my latest comment? 😊

Reply:
Reading the tests you linked I'm wondering how it even works since there's no `await` being used by `retry_any`. What do I miss?

Reply:
[The async `retry_any` does use `await`](https://github.com/jd/tenacity/pull/451/files#diff-e40ea0a413e972c47c884a384af80c3cbe4934358aa1d330447905320c8177d2R88-R100). This in combination with the `__ror__` and `__rand__` in the base class makes sure that we always use the appropriate any/all version.
- If we do `sync_strategy | sync_strategy` then it uses the regular `__and__` for the first strategy, so it results in using the sync `retry_any`.
- If we do `async_strategy | sync_strategy` then it uses the regular `__and__` for the first strategy, so it results in using the async `retry_any`.
- If we do `sync_strategy | async_strategy` then it uses the `__rand__` for the second strategy, as it's a subclass of the first one, so it results in using the async `retry_any`.
- If we do `async_strategy | async_strategy` then it uses the regular `__and__` for the first strategy, so it results in using the async `retry_any`.

I agree it does feel a bit odd, but I think this results in quite a good abstraction, as the setup in the base `retry_any`/`retry_all` allows other types of strategies (in the case of the changes here, async bases) to override this behaviour, while the original sync bases don't need to know about these implementations at all.

The tests I linked work because regardless of the types of strategies used, they are always triggered when using `AsyncRetrying(retry=retry_strategy)`, which requires doing so with `async for`, it's then the internals that make sure all of the callables and strategies are async and called as such.